cairo.html.markdown

```diff
@@ -0,0 +1,862 @@
+const TOKEN_TYPE_B = 2;
+
+// @dev Ensure the user's balances are much smaller than the pool's balance
+const POOL_UPPER_BOUND = 2 ** 30;
+const ACCOUNT_BALANCE_BOUND = 1073741; // (2 ** 30 / 1000)
+
+
+// STORAGE VARIABLES
+//
+// @dev A map from account and token type to corresponding balance
+@storage_var
+func account_balance(account_id: felt, token_type: felt) -> (balance: felt){}
+
+// @dev a map from token type to corresponding pool balance
+@storage_var
+func pool_balance(token_type: felt) -> (balance: felt) {}
+
+
+// GETTERS
+//
+// @dev returns account balance for a given token
+// @param account_id Account to be queried
+// @param token_type Token to be queried
+@view
+func get_account_token_balance{syscall_ptr: felt*, pedersen_ptr:
+  HashBuiltin*, range_check_ptr}(
+    account_id: felt, token_type: felt
+    ) -> (balance: felt) {
+    return account_balance.read(account_id, token_type);
+  }
+
+// @dev return the pool's balance
+// @param token_type Token type to get pool balance
+@view
+func get_pool_token_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*,
+  range_check_ptr}(
+    token_type: felt
+    ) -> (balance: felt) {
+    return pool_balance.read(token_type);
+  }
+
+
+// EXTERNALS
+//
+// @dev set pool balance for a given token
+// @param token_type Token whose balance is to be set
+// @param balance Amount to be set as balance
+@external
+func set_pool_token_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*,
+  range_check_ptr}(
+    token_type: felt, balance: felt
+    ) {
+    with_attr error_message("exceeds maximum allowed tokens!"){
+      assert_nn_le(balance, BALANCE_UPPER_BOUND - 1);
+    }
+
+  pool_balance.write(token_type, balance);
+  return ();
+  }
+
+// @dev add demo token to the given account
+// @param token_a_amount amount of token a to be added
+// @param token_b_amount amount of token b to be added
+@external
+func add_demo_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*,
+  range_check_ptr}(
+    token_a_amount: felt, token_b_amount: felt
+  ) {
+    alloc_locals;
+    let (account_id) = get_caller_address();
+
+    modify_account_balance(account_id=account_id, token_type=TOKEN_TYPE_A,
+      amount=token_a_amount);
+    modify_account_balance(account_id=account_id, token_type=TOKEN_TYPE_B,
+      amount=token_b_amount);
+
+    return ();
+  }
+
+// @dev intialize AMM
+// @param token_a amount of token a to be set in pool
+// @param token_b amount of token b to be set in pool
+@external
+func init_pool{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*,
+  range_check_ptr}(
+    token_a: felt, token_b: felt
+    ) {
+      with_attr error_message("exceeds maximum allowed tokens!"){
+        assert_nn_le(token_a, POOL_UPPER_BOUND - 1);
+        assert_nn_le(token_b, POOL_UPPER_BOUND - 1);
+      }
+
+      set_pool_token_balance(token_type=TOKEN_TYPE_A, balance=token_a);
+      set_pool_token_balance(token_type=TOKEN_TYPE_B, balance=token_b);
+
+      return ();
+  }
```

Comment:
The indentation for function code blocks seems to be pretty inconsistent throughout the document.